src/components/Popup/Popup.js

```diff
@@ -0,0 +1,167 @@
+        const { isOpened } = this.state;
+        if (isOpened && !prevState.isOpened) {
+            this.createPopperInstance();
+        }
+        if (!isOpened && prevState.isOpened) {
+            this.destroyPopperIstance();
+        }
+    }
+
+    componentWillUnmount() {
+        document.removeEventListener('mousedown', this.handleWindowClick);
+        document.removeEventListener('keydown', this.handleEscKeyPress, false);
+    }
+
+    setPopupVisibility = isOpened => {
+        if (isOpened !== this.state.isOpened) { // eslint-disable-line
+            this.setState({ isOpened });
+        }
+    };
+
+    getArgs() {
+        const { isOpened } = this.state;
+        const args = {
+            setPopupVisibility: this.setPopupVisibility,
+            isOpened,
+        };
+        if (this.anchorRef.current) {
+            const { width } = this.anchorRef.current.getBoundingClientRect();
+            args.anchorWidth = width;
+        }
+        return args;
+    }
+
+    handleWindowClick = event => {
+        if (!this.contentRef.current) return;
+        const wasPopupClicked = Popup.checkDOMPath(
+            event.target,
+            elem => elem.dataset.popup === 'true'
+        );
+        const wasOutsideClicked = this.contentRef.current.contains(event.target) === false;
+        const wasAnchorClicked =
+            this.anchorRef.current && this.anchorRef.current.contains(event.target) === true;
+        if (wasPopupClicked === false && wasOutsideClicked === true && wasAnchorClicked === false) {
+            this.close();
+        }
+    };
+
+    handleEscKeyPress = event => {
+        if (event.keyCode === 27) {
+            this.close();
+        }
+    };
+
+    close() {
+        const { isOpened } = this.state;
+        if (isOpened === true) {
+            this.setState({ isOpened: false });
+        }
+    }
+
+    createPopperInstance() {
+        if (!this.anchorRef.current || !this.contentRef.current) return;
+        const { placement } = this.props;
+        this.destroyPopperIstance();
+        this.popper = new PopperJS(this.anchorRef.current, this.contentRef.current, {
+            placement,
+        });
+    }
+
+    destroyPopperIstance() {
+        if (!this.popper) return;
+        this.popper.destroy();
+        this.popper = undefined;
+    }
+
+    renderAnchor() {
+        const { anchor } = this.props;
+        const anchorElem = anchor(this.getArgs());
+        return React.cloneElement(anchorElem, { ref: this.anchorRef });
+    }
+
+    renderContent() {
+        const { isOpened } = this.state;
+        if (!isOpened) return null;
+        const { content } = this.props;
+        const contentElem = content(this.getArgs());
+        const contentElemWithRef = React.cloneElement(contentElem, {
+            ref: this.contentRef,
+            'data-popup': 'true',
+        });
+        return ReactDOM.createPortal(contentElemWithRef, document.body);
+    }
+
+    render() {
+        return (
```

Comment:
or using render-props pattern, the `renderAnchor` and `renderContent` functions above could be removed and this render replaced with:
```
render() {
        const { children } = this.props;
        return children({
            ...this.getArgs(),
            anchorRef: this.anchorRef,
            contentRef: this.contentRef,
            contentProps: { 'data-popup': true },
        });
    }
```